### ISSUE TITLE
libjack object files included in libjack and libjackserver

### ISSUE THREAD

**Issue by danielhams** [jackaudio/jack1#12]
Overview:
-------------

While messing about with C++ bits in jack1, I noticed that static thread_local variables were being destroyed multiple times. I tracked it down to the object files being built and included twice.

The current automake files build versions of the following files and include them in both libjack.so and libjackserver.so:

../libjack/systemtest.c
../libjack/sanitycheck.c
../libjack/client.c
... etc

Notably it's the ones that appear in the jackd/Makefile.am file.

Problem Reproduction:
--------------------
Rather than pollute jackd with the bits for building C++, I've reproduced the issue using GCC's constructor and destructor annotations.

When dlopen and dlclose of a driver linked against libjackserver.so occurs, any methods that are annotated with constructor and destructor are called multiple times due to the symbols being included in both the libjack.so and libjackserver.so.

There is a branched version of jack1 that illustrates this problem here:

https://github.com/danielhams/jack1/tree/librarybuildissue

When running jackd, you will see the following in the output:

http://pastebin.com/KwLeJkRs

Note the calls to some_constructor and some_destructor on dlopen and dlclose in addition to the ones at jackd start and stop.

Proposed Resolution:
-----------------------------

If you switch to the branch:

https://github.com/danielhams/jack1/tree/librarybuildissuefix

You now get the following output:

http://pastebin.com/uSZw1Daf

The constructor and destructor have the correct lifecycle as they follow the main process lifetime.

Basically it's just moving those libjack objects into the libjack/Makefile.am and making the libjackserver and jackd link against the libjack to resolve those symbols.

<B>Note: the above branches are for demonstration purposes and include throw away code, I'll follow this comment with the appropriate pull request for a fix on the regular jack master branch.<B>

**Comment by pauldavisthefirst** [comment]
There seems to be some confusion here. 

libjack and libjackserver should never be linked against by the same process. libjackserver is specifically intended as a special version of libjack for use by the server (and also by programs that can actually become the server). 

I have reviewed the drivers, and they all use libjackserver only. jackd also ONLY links against libjackserver. 

Where and/or why do you believe that a process is being linked against both libraries?

**Comment by danielhams** [comment]
Wow, I thought this had gravestoned.

I'll have to scour a little back through that and the code to see what I
was getting at.

Off the top of my head it wasn't so much about what was being linked as the
fact that symbols were present in multiple places, and for some reason I
was seeing C++ things destroyed when they shouldn't be.

Give me a few months .-)

D

**Comment by pauldavisthefirst** [comment]
They *should* be in multiple places, but those places should never come together in a single process. That's the idea, at least.

**Comment by danielhams** [comment]
Evening Paul,

> They *should* be in multiple places

What is the rational for this? Surely shared libraries are a great thing to
use? The suggested resolution patch was about linking libjackserver.so
against libjack.so and having jackd use both. The motivation was from
skipping some lifecycle issues arising from the duplicated symbols - C++
static lifecycle was how I originally noticed it (IIRC, it's been a while).

Are there perhaps cache coherency issues or having the code in a single
blob dictating this?

The reason I ask is that if there's something I'm missing (eminently
probable) I'll skip on the deeper investigation .-)

D

**Comment by pauldavisthefirst** [comment]
Your thinking too hard. libjackserver exists to build JACK servers and their backends. libjack exists to build clients. They are not intended to both be used by a single process. libjackserver contains some special extra magic compared to libjack that allows a process linked against it to *become* a server (even if it is not jackd), and this extra magic requires the use of a different library. 

You should probably take the suggestion to jack-devel, where Stephane Letz (who originated the libjackserver idea) *might* comment on it.
